# Aviation Accidents Analysis

You are part of a consulting firm that is tasked to do an analysis of commercial and passenger jet airline safety. The client (an airline/airplane insurer) is interested in knowing what types of aircraft (makes/models) exhibit low rates of total destruction and low likelihood of fatal or serious passenger injuries in the event of an accident. They are also interested in any general variables/conditions that might be at play. Your analysis will be based off of aviation accident data accumulated from the years 1948-2023. 

Our client is only interested in airplane makes/models that are professional builds and could potentially still be active. Assume a max lifetime of 40 years for a make/model retirement and make sure to filter your data accordingly (i.e. from 1983 onwards). They would also like separate recommendations for small aircraft vs. larger passenger models. **In addition, make sure that claims that you make are statistically robust and that you have enough samples when making comparisons between groups.**


In this summative assessment you will demonstrate your ability to:
- **Use Pandas to load, inspect, and clean the dataset appropriately.**
- **Transform relevant columns to create measures that address the problem at hand.**
- conduct EDA: visualization and statistical measures to systematically understand the structure of the data
- recommend a set of airplanes and makes conforming to the client's request and identify at least *two* factors contributing to airplane safety. You must provide supporting evidence (visuals, summary statistics, tables) for each claim you make.

### Make relevant library imports

In [40]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

import datetime as datetime
import warnings 
warnings.filterwarnings('ignore')

## Data Loading and Inspection

### Load in data from the relevant directory and inspect the dataframe.
- inspect NaNs, datatypes, and summary statistics

In [41]:
# Load the data set 
df = pd.read_csv('AviationData.csv', index_col=0, encoding='latin1')

# Check shape
df.shape
print(f'My dataset has {df.shape[0]} rows and {df.shape[1]} columns')

My dataset has 88889 rows and 30 columns


In [42]:
# Set display options for better readability
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 100)

In [43]:
#Check top 5 rows
df.head()

,Investigation.Type,Accident.Number,Event.Date,Location,Country,Latitude,Longitude,Airport.Code,Airport.Name,Injury.Severity,Aircraft.damage,Aircraft.Category,Registration.Number,Make,Model,Amateur.Built,Number.of.Engines,Engine.Type,FAR.Description,Schedule,Purpose.of.flight,Air.carrier,Total.Fatal.Injuries,Total.Serious.Injuries,Total.Minor.Injuries,Total.Uninjured,Weather.Condition,Broad.phase.of.flight,Report.Status,Publication.Date
Event.Id,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
20001218X45444,Accident,SEA87LA080,1948-10-24,"MOOSE CREEK, ID",United States,NaN,NaN,NaN,NaN,Fatal(2),Destroyed,NaN,NC6404,Stinson,108-3,No,1.0,Reciprocating,NaN,NaN,Personal,NaN,2.0,0.0,0.0,0.0,UNK,Cruise,Probable Cause,NaN
20001218X45447,Accident,LAX94LA336,1962-07-19,"BRIDGEPORT, CA",United States,NaN,NaN,NaN,NaN,Fatal(4),Destroyed,NaN,N5069P,Piper,PA24-180,No,1.0,Reciprocating,NaN,NaN,Personal,NaN,4.0,0.0,0.0,0.0,UNK,Unknown,Probable Cause,19-09-1996
20061025X01555,Accident,NYC07LA005,1974-08-30,"Saltville, VA",United States,36.922223,-81.878056,NaN,NaN,Fatal(3),Destroyed,NaN,N5142R,Cessna,172M,No,1.0,Reciprocating,NaN,NaN,Personal,NaN,3.0,NaN,NaN,NaN,IMC,Cruise,Probable Cause,26-02-2007
20001218X45448,Accident,LAX96LA321,1977-06-19,"EUREKA, CA",United States,NaN,NaN,NaN,NaN,Fatal(2),Destroyed,NaN,N1168J,Rockwell,112,No,1.0,Reciprocating,NaN,NaN,Personal,NaN,2.0,0.0,0.0,0.0,IMC,Cruise,Probable Cause,12-09-2000
20041105X01764,Accident,CHI79FA064,1979-08-02,"Canton, OH",United States,NaN,NaN,NaN,NaN,Fatal(1),Destroyed,NaN,N15NY,Cessna,501,No,NaN,NaN,NaN,NaN,Personal,NaN,1.0,2.0,NaN,0.0,VMC,Approach,Probable Cause,16-04-1980


In [44]:
#check bottom 5
df.tail()

,Investigation.Type,Accident.Number,Event.Date,Location,Country,Latitude,Longitude,Airport.Code,Airport.Name,Injury.Severity,Aircraft.damage,Aircraft.Category,Registration.Number,Make,Model,Amateur.Built,Number.of.Engines,Engine.Type,FAR.Description,Schedule,Purpose.of.flight,Air.carrier,Total.Fatal.Injuries,Total.Serious.Injuries,Total.Minor.Injuries,Total.Uninjured,Weather.Condition,Broad.phase.of.flight,Report.Status,Publication.Date
Event.Id,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
20221227106491,Accident,ERA23LA093,2022-12-26,"Annapolis, MD",United States,NaN,NaN,NaN,NaN,Minor,NaN,NaN,N1867H,PIPER,PA-28-151,No,NaN,NaN,091,NaN,Personal,NaN,0.0,1.0,0.0,0.0,NaN,NaN,NaN,29-12-2022
20221227106494,Accident,ERA23LA095,2022-12-26,"Hampton, NH",United States,NaN,NaN,NaN,NaN,NaN,NaN,NaN,N2895Z,BELLANCA,7ECA,No,NaN,NaN,NaN,NaN,NaN,NaN,0.0,0.0,0.0,0.0,NaN,NaN,NaN,NaN
20221227106497,Accident,WPR23LA075,2022-12-26,"Payson, AZ",United States,341525N,1112021W,PAN,PAYSON,Non-Fatal,Substantial,Airplane,N749PJ,AMERICAN CHAMPION AIRCRAFT,8GCBC,No,1.0,NaN,091,NaN,Personal,NaN,0.0,0.0,0.0,1.0,VMC,NaN,NaN,27-12-2022
20221227106498,Accident,WPR23LA076,2022-12-26,"Morgan, UT",United States,NaN,NaN,NaN,NaN,NaN,NaN,NaN,N210CU,CESSNA,210N,No,NaN,NaN,091,NaN,Personal,MC CESSNA 210N LLC,0.0,0.0,0.0,0.0,NaN,NaN,NaN,NaN
20221230106513,Accident,ERA23LA097,2022-12-29,"Athens, GA",United States,NaN,NaN,NaN,NaN,Minor,NaN,NaN,N9026P,PIPER,PA-24-260,No,NaN,NaN,091,NaN,Personal,NaN,0.0,1.0,0.0,1.0,NaN,NaN,NaN,30-12-2022


In [45]:
df.sample(5)

,Investigation.Type,Accident.Number,Event.Date,Location,Country,Latitude,Longitude,Airport.Code,Airport.Name,Injury.Severity,Aircraft.damage,Aircraft.Category,Registration.Number,Make,Model,Amateur.Built,Number.of.Engines,Engine.Type,FAR.Description,Schedule,Purpose.of.flight,Air.carrier,Total.Fatal.Injuries,Total.Serious.Injuries,Total.Minor.Injuries,Total.Uninjured,Weather.Condition,Broad.phase.of.flight,Report.Status,Publication.Date
Event.Id,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
20001208X05807,Accident,LAX96LA192,1996-05-07,"RENO, NV",United States,NaN,NaN,4SD,RENO/STEAD,Non-Fatal,Substantial,NaN,N5394K,Cessna,172P,No,1.0,Reciprocating,NaN,NaN,Personal,NaN,0.0,0.0,0.0,2.0,VMC,Approach,Probable Cause,25-08-1997
20001208X08170,Accident,LAX97LA206,1997-06-10,"HERON, CA",United States,NaN,NaN,NaN,NaN,Non-Fatal,Destroyed,NaN,N888SJ,Bell,UH-1B,No,1.0,Turbo Shaft,NaN,UNK,Aerial Application,NaN,0.0,0.0,0.0,1.0,VMC,Maneuvering,Probable Cause,21-05-1998
20150909X33413,Accident,CEN15LA403,2015-08-16,"Big Piney, WY",United States,043935N,1093932W,NaN,NaN,Non-Fatal,Substantial,Airplane,XBYUH,CESSNA,182,No,1.0,Reciprocating,091,NaN,Personal,NaN,0.0,2.0,0.0,0.0,VMC,NaN,The pilot's failure to maintain adequate altit...,25-09-2020
20010105X00033,Accident,FTW01LA042,2000-12-22,"MESQUITE, TX",United States,NaN,NaN,HQZ,MESQUITE METRO,Non-Fatal,Substantial,NaN,N7158M,Robinson,R-22B2,No,1.0,Reciprocating,NaN,NaN,Instructional,NaN,0.0,1.0,1.0,0.0,VMC,Maneuvering,Probable Cause,19-04-2001
20001214X37717,Accident,ATL85MA282B,1985-09-22,"AUBURN, AL",United States,NaN,NaN,AUO,AUBURN-OPELIKA,Fatal(2),Destroyed,NaN,NONE,Challenger,II,Yes,1.0,Reciprocating,NaN,NaN,Ferry,NaN,2.0,5.0,1.0,0.0,VMC,Approach,Probable Cause,NaN


Observation: the dataset is uniform form top to bottom

In [46]:
#Check columns
df.columns

Index(['Investigation.Type', 'Accident.Number', 'Event.Date', 'Location',
       'Country', 'Latitude', 'Longitude', 'Airport.Code', 'Airport.Name',
       'Injury.Severity', 'Aircraft.damage', 'Aircraft.Category',
       'Registration.Number', 'Make', 'Model', 'Amateur.Built',
       'Number.of.Engines', 'Engine.Type', 'FAR.Description', 'Schedule',
       'Purpose.of.flight', 'Air.carrier', 'Total.Fatal.Injuries',
       'Total.Serious.Injuries', 'Total.Minor.Injuries', 'Total.Uninjured',
       'Weather.Condition', 'Broad.phase.of.flight', 'Report.Status',
       'Publication.Date'],
      dtype='object')

Observation: Theres are capital letter and fullstops so we will have to change that to lowercase and underscores

COLUMN UNDERSTANDING
- Event.Id — unique identifier for the accident event
- Accident.Number — case number
- Investigation.Type — whether it was an Accident or Incident
- Report.Status — status of theinvestigation report
- Publication.Date — date the report was officially published
- Event.Date — date the accident occurred
- Location — city or town where the accident happened
- Country — country where the accident happened
- Latitude / Longitude — GPS coordinates of the accident location
- Airport.Code — code of the airport
- Airport.Name — name of the airport
- Make — manufacturer of the aircraft (e.g. Cessna, Boeing, Piper)
- Model — specific model of the aircraft (e.g. 172, 737)
-Registration.Number — number used to identify the specific aircraft
- Aircraft.Category — type of aircraft (e.g. Airplane, Helicopter, Glider)
- Aircraft.damage — damage to the aircraft (Minor, Substantial, Destroyed)
- Amateur.Built — whether the aircraft was built by an amateur (Yes or No)
- Number.of.Engines — how many engines the aircraft has
- Engine.Type — type of engine (e.g. Reciprocating, Turbo Fan, Turbo Prop)
- FAR.Description — FAA regulatory category under which the flight operated
- Air.carrier — name of the airline or commercial operator
- Schedule — whether the flight was scheduled or unscheduled
- Purpose.of.flight — reason for the flight (e.g. Personal, Instructional, Business)
- Broad.phase.of.flight — stage of flight when the accident occurred (e.g. Takeoff, Cruise, Landing)
- Weather.Condition — visibility conditions at the time of the accident 
- Injury.Severity — categorical summary of the worst injury in the accident (Fatal, Serious, Minor, None)
- Total.Fatal.Injuries — number of people killed
- Total.Serious.Injuries — number of people seriously injured
- Total.Minor.Injuries — number of people with minor injuries
- Total.Uninjured — number of people on board with no injuries

In [47]:
#Check datatypes
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 88889 entries, 20001218X45444 to 20221230106513
Data columns (total 30 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Investigation.Type      88889 non-null  object 
 1   Accident.Number         88889 non-null  object 
 2   Event.Date              88889 non-null  object 
 3   Location                88837 non-null  object 
 4   Country                 88663 non-null  object 
 5   Latitude                34382 non-null  object 
 6   Longitude               34373 non-null  object 
 7   Airport.Code            50132 non-null  object 
 8   Airport.Name            52704 non-null  object 
 9   Injury.Severity         87889 non-null  object 
 10  Aircraft.damage         85695 non-null  object 
 11  Aircraft.Category       32287 non-null  object 
 12  Registration.Number     87507 non-null  object 
 13  Make                    88826 non-null  object 
 14  Model                

Observation: 25 columns contain categorical data and 5 columns contain numerical. However we should set datatype of publication data and event date to datetime

In [48]:
#Check unique
for coln in df.columns:
    unique_value = df[coln].unique()
    print(f'{coln}\n {unique_value}\n')

Investigation.Type
 ['Accident' 'Incident']

Accident.Number
 ['SEA87LA080' 'LAX94LA336' 'NYC07LA005' ... 'WPR23LA075' 'WPR23LA076'
 'ERA23LA097']

Event.Date
 ['1948-10-24' '1962-07-19' '1974-08-30' ... '2022-12-22' '2022-12-26'
 '2022-12-29']

Location
 ['MOOSE CREEK, ID' 'BRIDGEPORT, CA' 'Saltville, VA' ... 'San Manual, AZ'
 'Auburn Hills, MI' 'Brasnorte, ']

Country
 ['United States' nan 'GULF OF MEXICO' 'Puerto Rico' 'ATLANTIC OCEAN'
 'HIGH ISLAND' 'Bahamas' 'MISSING' 'Pakistan' 'Angola' 'Germany'
 'Korea, Republic Of' 'Martinique' 'American Samoa' 'PACIFIC OCEAN'
 'Canada' 'Bolivia' 'Mexico' 'Dominica' 'Netherlands Antilles' 'Iceland'
 'Greece' 'Guam' 'Australia' 'CARIBBEAN SEA' 'West Indies' 'Japan'
 'Philippines' 'Venezuela' 'Bermuda' 'San Juan Islands' 'Colombia'
 'El Salvador' 'United Kingdom' 'British Virgin Islands' 'Netherlands'
 'Costa Rica' 'Mozambique' 'Jamaica' 'Panama' 'Guyana' 'Norway'
 'Hong Kong' 'Portugal' 'Malaysia' 'Turks And Caicos Islands'
 'Northern Mariana I

In [49]:
# check statistic
df.describe().T

,count,mean,std,min,25%,50%,75%,max
Number.of.Engines,82805.0,1.146585,0.446510,0.0,1.0,1.0,1.0,8.0
Total.Fatal.Injuries,77488.0,0.647855,5.485960,0.0,0.0,0.0,0.0,349.0
Total.Serious.Injuries,76379.0,0.279881,1.544084,0.0,0.0,0.0,0.0,161.0
Total.Minor.Injuries,76956.0,0.357061,2.235625,0.0,0.0,0.0,0.0,380.0
Total.Uninjured,82977.0,5.325440,27.913634,0.0,0.0,1.0,2.0,699.0


In [50]:
#check categorical columns
df.describe(include='O').T

,count,unique,top,freq
Investigation.Type,88889,2,Accident,85015
Accident.Number,88889,88863,CEN22LA149,2
Event.Date,88889,14782,1984-06-30,25
Location,88837,27758,"ANCHORAGE, AK",434
Country,88663,219,United States,82248
Latitude,34382,25592,332739N,19
Longitude,34373,27156,0112457W,24
Airport.Code,50132,10374,NONE,1488
Airport.Name,52704,24870,Private,240
Injury.Severity,87889,109,Non-Fatal,67357


In [51]:
#Check for duplicates
print(f"There are {df.duplicated().sum()} duplicates in the dataset")

There are 0 duplicates in the dataset


In [52]:
#check null values
df.isna().sum()

Investigation.Type            0
Accident.Number               0
Event.Date                    0
Location                     52
Country                     226
Latitude                  54507
Longitude                 54516
Airport.Code              38757
Airport.Name              36185
Injury.Severity            1000
Aircraft.damage            3194
Aircraft.Category         56602
Registration.Number        1382
Make                         63
Model                        92
Amateur.Built               102
Number.of.Engines          6084
Engine.Type                7096
FAR.Description           56866
Schedule                  76307
Purpose.of.flight          6192
Air.carrier               72241
Total.Fatal.Injuries      11401
Total.Serious.Injuries    12510
Total.Minor.Injuries      11933
Total.Uninjured            5912
Weather.Condition          4492
Broad.phase.of.flight     27165
Report.Status              6384
Publication.Date          13771
dtype: int64

In [53]:
# check null values percentage
missing_values_percentage = (df.isna().sum()/len(df)*100).sort_values(ascending=False)
missing_values_percentage

Schedule                  85.845268
Air.carrier               81.271023
FAR.Description           63.974170
Aircraft.Category         63.677170
Longitude                 61.330423
Latitude                  61.320298
Airport.Code              43.601570
Airport.Name              40.708074
Broad.phase.of.flight     30.560587
Publication.Date          15.492356
Total.Serious.Injuries    14.073732
Total.Minor.Injuries      13.424608
Total.Fatal.Injuries      12.826109
Engine.Type                7.982990
Report.Status              7.181991
Purpose.of.flight          6.965991
Number.of.Engines          6.844491
Total.Uninjured            6.650992
Weather.Condition          5.053494
Aircraft.damage            3.593246
Registration.Number        1.554748
Injury.Severity            1.124999
Country                    0.254250
Amateur.Built              0.114750
Model                      0.103500
Make                       0.070875
Location                   0.058500
Accident.Number            0

## Data Cleaning

### Filtering aircrafts and events

We want to filter the dataset to include aircraft that the client is interested in an analysis of:
- inspect relevant columns
- figure out any reasonable imputations
- filter the dataset

In [54]:
#Make a copy
airline_df = df.copy(deep=True)


In [55]:
#Remove whitespaces and make lowercase letter 
airline_df.columns = airline_df.columns.str.replace(".", "_", regex=False).str.lower().str.strip()

Doing this makes all column names uniform

In [56]:
#Change event date and publication date to datetime datatype
airline_df['publication_date']= pd.to_datetime(airline_df['publication_date'], format= '%d-%m-%Y')
airline_df['event_date']= pd.to_datetime(airline_df['event_date'], format= '%Y-%m-%d')

#Filter event date from 1983 onwards
airline_df = airline_df[airline_df['event_date'].dt.year >= 1983]

Changed the data type of eventdate and publication date to datetime and filtered event date to 1983 onwards

In [57]:
#Filter column inestigation type to have only rows with accident
airline_df= airline_df[airline_df['investigation_type'] == 'Accident']

#Filter amateur aircrafts to have only professional build
airline_df= airline_df[airline_df['amateur_built'] == 'No']

Filtering investigation type to accident only and amateur build to 'no' as specified for analysis

### Cleaning and constructing Key Measurables

Injuries and robustness to destruction are a key interest point for the client. Clean and impute relevant columns and then create derived fields that best quantifies what the client wishes to track. **Use commenting or markdown to explain any cleaning assumptions as well as any derived columns you create.**

**Construct metric for fatal/serious injuries**

*Hint:* Estimate the total number of passengers on each flight. The likelihood of serious / fatal injury can be estimated as a fraction from this.

In [58]:
#Filling in null values in numerical columns of injuries
airline_df= airline_df.fillna({'total_fatal_injuries':0,
                                'total_serious_injuries':0, 
                                'total_minor_injuries':0, 
                                'total_uninjured':0})

Observation: Filling injury and people whosurvived with 0 because its the mode for all columns and doing this so that calculations can be done without null


In [59]:
#Calculating total passengers onboard
airline_df['total_passengers'] = (airline_df['total_fatal_injuries'] + 
                                  airline_df['total_serious_injuries'] + 
                                  airline_df['total_minor_injuries'] +
                                   airline_df['total_uninjured'])


Observation: Ive created a column of total passengers

In [60]:
#Fraction of serious and fatal injuries on each flight
airline_df['serious_and_fatal_fraction'] = (airline_df['total_fatal_injuries'] + airline_df['total_serious_injuries']) / airline_df['total_passengers']


Observation: Ive created column of fraction of fatal and serios injuries 

**Aircraft.Damage**
- identify and execute any cleaning tasks
- construct a derived column tracking whether an aircraft was destroyed or not.

In [61]:
#Before deriving column, impute the null vales with mode
airline_df['aircraft_damage'] = airline_df['aircraft_damage'].fillna(airline_df['aircraft_damage'].mode()[0])

#Create column
airline_df['aircraft_destroyed'] = airline_df['aircraft_damage'] == 'Destroyed'

Observation: Ive created a column with bool values to tell whther or not the aircraft was destroyed or not

### Investigate the *Make* column
- Identify cleaning tasks here
- List cleaning tasks clearly in markdown
- Execute the cleaning tasks
- For your analysis, keep Makes with a reasonable number (you can put the threshold at 50 though lower could work as well)

In [62]:
#Identifying what should be cleaned in make column
unique_value = airline_df['make'].unique()

print(f'{'make'}\n {unique_value}\n')

print(f'Missing values : {airline_df['make'].isna().sum()}')

make
 ['Piccard' 'Cessna' 'Piper' ... 'JAMES R DERNOVSEK' 'ORLICAN S R O'
 'ROYSE RALPH L']

Missing values : 20


List of cleaning: 
1. Theres inconsistent casing of values
2. Same names appear in different spellings
3. There are 20 missing values

In [63]:
#Cleaning the make column
airline_df['make'] = airline_df['make'].str.strip().str.capitalize()

#Drop the null values 
airline_df = airline_df.dropna(subset=['make'])

#Replace the names with multiple spellings as one
replacements = {
    'Piper aircraft': 'Piper',
    'Beech': 'Beechcraft',
    'Boeing co': 'Boeing',
    'Mcdonnell douglas aircraft co': 'Mcdonnell douglas',
    'Grumman american': 'Grumman',
    'Citrus design corp': 'Citrus',
    'Robinson helicopter company': 'Robinson helicopter',
    'Air tractor inc':'Air tractor'
    }

airline_df['make'] = airline_df['make'].replace(replacements)

In [64]:
# Keep make threshold to 50
make_count = airline_df['make'].value_counts()
threshold_count = make_count[make_count >= 50].index
airline_df = airline_df[airline_df['make'].isin(threshold_count)]

Observation: Ive made the column make to have only incidents above 50

### Inspect Model column
- Get rid of any NaNs.
- Inspect the column and counts for each model/make. Are model labels unique to each make?
- If not, create a derived column that is a unique identifier for a given plane type.

In [65]:
#Remove null
airline_df = airline_df.dropna(subset=['model'])

In [66]:
#Make and model columns
#Inspect count of column and counts
airline_df[['model','make']].value_counts()

model      make   
152        Cessna     2215
172        Cessna     1640
172N       Cessna     1091
PA-28-140  Piper       862
172M       Cessna      754
                      ... 
C-305A     Cessna        1
C-310B     Cessna        1
C-310Q     Cessna        1
C-310R     Cessna        1
ZL-60      Cameron       1
Name: count, Length: 5942, dtype: int64

Observation: Many model labels are shared across different manufacturers but model labels alone are not unique identifiers for a specific aircraft type

In [67]:
#combine make and model column as a new column
airline_df['model_make'] = airline_df['make'] + '  ' + airline_df['model']


### Cleaning other columns
- there are other columns containing data that might be related to the outcome of an accident. We list a few here:
- Engine.Type
- Weather.Condition
- Number.of.Engines
- Purpose.of.flight
- Broad.phase.of.flight

Inspect and identify potential cleaning tasks in each of the above columns. Execute those cleaning tasks. 

**Note**: You do not necessarily need to impute or drop NaNs here.

In [68]:
#Check value count of columns
print(airline_df[['number_of_engines']].value_counts(),'\n',
      airline_df[['engine_type']].value_counts(),'\n',
      airline_df[['purpose_of_flight']].value_counts(),'\n',
      airline_df[['weather_condition']].value_counts(),'\n',
      airline_df[['broad_phase_of_flight']].value_counts())


number_of_engines
1.0                  54004
2.0                   8023
0.0                    725
4.0                    200
3.0                    176
8.0                      1
Name: count, dtype: int64 
 engine_type  
Reciprocating    54423
Turbo Shaft       2896
Turbo Prop        2451
Unknown           1404
Turbo Fan         1104
Turbo Jet          312
LR                   1
NONE                 1
UNK                  1
Name: count, dtype: int64 
 purpose_of_flight        
Personal                     36658
Instructional                 9280
Unknown                       4331
Aerial Application            4128
Business                      3325
Positioning                   1368
Other Work Use                1041
Aerial Observation             705
Public Aircraft                624
Ferry                          619
Executive/corporate            391
Skydiving                      174
Flight Test                    147
Banner Tow                      97
External Load              

In [69]:
#Replacing engine number that have 0 with null value becase an engine cant have 0 engine
airline_df['number_of_engines'] = airline_df['number_of_engines'].replace(0.0, np.nan)

In [70]:
# Other value basivally means the phase of flight was unknown
airline_df['broad_phase_of_flight'] = airline_df['broad_phase_of_flight'].replace('Other','Unknown')


In [71]:
#making all values uppercase
airline_df['weather_condition'] = airline_df['weather_condition'].str.strip().str.upper()

Cleaning: 
- Engine type null be filled with unknown 
- Purpose of flight null be filled with unknown
- weather condition fill with mode
- Broad phase of flight has unknown and other which mean the same thing, so all will be pit to unknown


In [72]:
airline_df= airline_df.fillna({'engine_type':'Unknown',
                               'purpose_of_flight':'Unknown',
                               'broad_phase_of_flight':(airline_df['broad_phase_of_flight'].mode()[0]),
                               'weather_condition':(airline_df['weather_condition'].mode()[0]),
                               'number_of_engines':(airline_df['number_of_engines'].mode()[0])})

In [73]:
#Checking the value count after cleaning
print(airline_df[['number_of_engines']].value_counts(),'\n',
      airline_df[['engine_type']].value_counts(),'\n',
      airline_df[['purpose_of_flight']].value_counts(),'\n',
      airline_df[['weather_condition']].value_counts(),'\n',
      airline_df[['broad_phase_of_flight']].value_counts())

number_of_engines
1.0                  58488
2.0                   8023
4.0                    200
3.0                    176
8.0                      1
Name: count, dtype: int64 
 engine_type  
Reciprocating    54423
Unknown           5699
Turbo Shaft       2896
Turbo Prop        2451
Turbo Fan         1104
Turbo Jet          312
LR                   1
NONE                 1
UNK                  1
Name: count, dtype: int64 
 purpose_of_flight        
Personal                     36658
Instructional                 9280
Unknown                       7934
Aerial Application            4128
Business                      3325
Positioning                   1368
Other Work Use                1041
Aerial Observation             705
Public Aircraft                624
Ferry                          619
Executive/corporate            391
Skydiving                      174
Flight Test                    147
Banner Tow                      97
External Load                   85
Public Aircraft - F

In [74]:
airline_df[['number_of_engines','engine_type', 'broad_phase_of_flight','weather_condition','purpose_of_flight']].isna().sum()

number_of_engines        0
engine_type              0
broad_phase_of_flight    0
weather_condition        0
purpose_of_flight        0
dtype: int64

### Column Removal
- inspect the dataframe and drop any columns that have too many NaNs

In [75]:
#Insepect column null
airline_df.isna().sum().sort_values(ascending=False)/len(airline_df)*100

schedule                      87.317606
air_carrier                   82.919208
far_description               70.263724
aircraft_category             70.111231
longitude                     62.609138
latitude                      62.601662
airport_code                  43.240940
airport_name                  40.569310
publication_date              17.203385
report_status                  5.981641
registration_number            1.318622
serious_and_fatal_fraction     0.533728
injury_severity                0.400670
country                        0.264621
location                       0.053821
weather_condition              0.000000
total_uninjured                0.000000
investigation_type             0.000000
broad_phase_of_flight          0.000000
total_passengers               0.000000
total_serious_injuries         0.000000
aircraft_destroyed             0.000000
total_minor_injuries           0.000000
engine_type                    0.000000
total_fatal_injuries           0.000000


In [76]:
#Drop irrelevant columns
airline_df =airline_df.drop(['schedule',
                 'latitude',
                 'longitude',
                 'air_carrier',
                 'far_description',
                 'airport_code',
                 'airport_name'], axis=1)

### Save DataFrame to csv
- its generally useful to save data to file/server after its in a sufficiently cleaned or intermediate state
- the data can then be loaded directly in another notebook for further analysis
- this helps keep your notebooks and workflow readable, clean and modularized

In [77]:
#Saving the dataframe to use for EDA
airline_df.to_csv('Aviation_cleaned.csv', index=None)